From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: pthier <pthier@chromium.org>
Date: Tue, 25 Apr 2023 11:00:10 +0200
Subject: Handle empty ranges in unicode sets

If a unicode set operation contains only an empty range, we generated a
set expression without operands. However the expression should match
nothing, so add the special operand not matching anything instead.

Bug: chromium:1437346
Change-Id: I8dd58884aaf6915277c80effbda43ea715049146
Reviewed-on: https://chromium-review.googlesource.com/c/v8/v8/+/4474547
Commit-Queue: Patrick Thier <pthier@chromium.org>
Reviewed-by: Jakob Linke <jgruber@chromium.org>
Cr-Commit-Position: refs/heads/main@{#87257}

diff --git a/src/regexp/regexp-parser.cc b/src/regexp/regexp-parser.cc
index f420b2eec42d413941751e0d2831689fa820e260..0ba51712c8efb758d5081e790d52eb16a01b2366 100644
--- a/src/regexp/regexp-parser.cc
+++ b/src/regexp/regexp-parser.cc
@@ -2792,6 +2792,14 @@ RegExpTree* RegExpParserImpl<CharT>::ParseClassUnion(
     return ReportError(RegExpError::kNegatedCharacterClassWithStrings);
   }
 
+  if (operands->is_empty()) {
+    // Return empty expression if no operands were added (e.g. [\P{Any}]
+    // produces an empty range).
+    DCHECK(ranges->is_empty());
+    DCHECK(strings->empty());
+    return RegExpClassSetExpression::Empty(zone(), is_negated);
+  }
+
   return zone()->template New<RegExpClassSetExpression>(
       RegExpClassSetExpression::OperationType::kUnion, is_negated,
       may_contain_strings, operands);
diff --git a/test/mjsunit/regress/regress-crbug-1437346.js b/test/mjsunit/regress/regress-crbug-1437346.js
new file mode 100644
index 0000000000000000000000000000000000000000..d94789c0e5fdf4c5e30ac39e31a9cca62a731657
--- /dev/null
+++ b/test/mjsunit/regress/regress-crbug-1437346.js
@@ -0,0 +1,5 @@
+// Copyright 2023 the V8 project authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+assertFalse(/[\P{Any}]/v.test());
